.changelog/10964.txt

```diff
@@ -0,0 +1,3 @@
+```release-note:bug
+tls: use intermediate certificates when verifying TLS certificates.
```

Comment:
I would perhaps fine tune this changelog a bit, since to be fair, we do use intermediate certificates when verifying TLS, by setting: https://github.com/hashicorp/consul/blob/ea451971d4a5c4eb88808eba5e95742e7778a1fa/tlsutil/config.go#L523-L525
That ends up [here](https://github.com/golang/go/blob/3920d6f2085559ae262d651b00bf8b29f953580a/src/crypto/tls/handshake_server.go#L813-L829) in the Go `tls` package.

It is just this particular verification that does not use the client's intermediate CAs.

Reply:
> // RequireAndVerifyClientCert indicates that a client certificate should be requested
	// during the handshake, and that at least one valid certificate is required
	// to be sent by the client.

Reply:
During a TLS connection, the Go stdlib will end up [here](https://github.com/golang/go/blob/3920d6f2085559ae262d651b00bf8b29f953580a/src/crypto/tls/handshake_server.go#L813-L829) in `processCertsFromClient`:

Copying the code into the comment:
```
	if c.config.ClientAuth >= VerifyClientCertIfGiven && len(certs) > 0 {
		opts := x509.VerifyOptions{
			Roots:         c.config.ClientCAs,
			CurrentTime:   c.config.time(),
			Intermediates: x509.NewCertPool(),
			KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
		}

		for _, cert := range certs[1:] {
			opts.Intermediates.AddCert(cert)
		}

		chains, err := certs[0].Verify(opts)
		if err != nil {
			c.sendAlert(alertBadCertificate)
			return errors.New("tls: failed to verify client certificate: " + err.Error())
		}
```

So if `c.config.ClientAuth >= VerifyClientCertIfGiven && len(certs) > 0` is true, the `tls.Conn` will validate the client's certificate by adding the Intermediates like this PR does.

`RequireAndVerifyClientCert` is greater than `VerifyClientCertIfGiven`, so we should be falling into this statement.

Reply:
Thanks for clarifying!

Reply:
So that means that when the old code created a TLS connection as the server, it properly checked the intermediate certs. But when it was acting as the client, it wouldn't.

Reply:
I only looked at the code around verifying client certificates, but I would not be surprised if the Go stdlib did the same thing when acting as a client validating a server's certificates.